File: bindings/haskell/src/KeyNub/LicDongle.hs
{-# LANGUAGE ScopedTypeVariables #-}
-- | KeyNub License Dongle for Haskell.
--
-- Calls the SDK's flat companion API (@keynub_licdongle_flat@) through
-- function pointers resolved at run time, so nothing is linked at build time
-- and the package is pure Haskell. Handles are 'Dongle' values; byte data is
-- 'ByteString'; every failure is a 'LicDongleError' exception carrying the
-- library's status code, the operation and its diagnostic detail.
--
-- @
-- import KeyNub.LicDongle
--
-- main = withDongle open $ \\d -> do
--   _ <- verifyGenuine d                 -- throws unless genuine
--   secret <- withSession d $ appDecrypt d sealed   -- build the licence check on this
--   ...
-- @
module KeyNub.LicDongle
  ( -- * The native library
    libraryVersion
  , libraryPath
  , loadedLibraryPath
  , setLibraryPath
  , statusText
    -- * Failures
  , Status(..)
  , statusFromCode
  , LicDongleError(..)
  , LibraryError(..)
    -- * Dongles
  , Device(..)
  , devices
  , Dongle
  , open
  , openSerial
  , openPath
  , close
  , withDongle
  , serial
  , Info(..)
  , info
  , lastErrorDetail
    -- * Authenticity
  , GenuineResult(..)
  , verifyGenuine
  , isGenuine
  , setTrustRoot
    -- * Session
  , sessionOpen
  , sessionClose
  , withSession
  , authorizeWrite
  , rotateWriteKey
    -- * Records
  , RecordInfo(..)
  , records
  , readRecord
  , writeRecord
  , eraseRecord
  , eraseAllRecords
    -- * Counters
  , readCounter
  , incrementCounter
    -- * Application data
  , Scope(..)
  , appEncrypt
  , appDecrypt
  ) where

import Control.Exception (Exception(..), SomeException, bracket, bracket_, catch, throwIO)
import Control.Monad (forM, when)
import Data.Bits ((.&.))
import Data.ByteString (ByteString)
import qualified Data.ByteString as BS
import Data.Word (Word8)
import Foreign.C.String (CString, peekCString, withCString)
import Foreign.C.Types (CInt(..))
import Foreign.Marshal.Alloc (alloca, allocaBytes)
import Foreign.Marshal.Array (advancePtr, allocaArray, peekArray)
import Foreign.Ptr (Ptr, castPtr, nullPtr)
import Foreign.Storable (peek, poke, pokeByteOff)

import KeyNub.LicDongle.Library

-- | The library's status codes, by name.
data Status
  = InvalidArg | NoDevice | AccessDenied | IOFailure | Timeout | ProtocolError
  | NotGenuine | CertInvalid | SessionExpired | TagMismatch | Range | StorageFull
  | Busy | NotFound | AuthRequired | FirmwareIncompatible | SdkTooOld | Cancelled
  | NotImplemented | Internal
  | Unknown Int   -- ^ a code this binding does not know
  deriving (Eq, Show)

-- | The 'Status' for a raw status code.
statusFromCode :: Int -> Status
statusFromCode c = case c of
  -1 -> InvalidArg;  -2 -> NoDevice;      -3 -> AccessDenied;   -4 -> IOFailure
  -5 -> Timeout;     -6 -> ProtocolError; -7 -> NotGenuine;     -8 -> CertInvalid
  -9 -> SessionExpired; -10 -> TagMismatch; -11 -> Range;       -12 -> StorageFull
  -13 -> Busy;       -14 -> NotFound;     -15 -> AuthRequired;  -16 -> FirmwareIncompatible
  -17 -> SdkTooOld;  -18 -> Cancelled;    -19 -> NotImplemented; -20 -> Internal
  _ -> Unknown c

-- | A failed call: the status, its raw code, the C function that failed and
-- the library's diagnostic detail for it (often empty).
data LicDongleError = LicDongleError
  { errorStatus    :: Status
  , errorCode      :: Int
  , errorOperation :: String
  , errorDetail    :: String
  } deriving (Eq, Show)

instance Exception LicDongleError where
  displayException e =
    errorOperation e ++ ": " ++ show (errorStatus e) ++ " (" ++ show (errorCode e) ++ ")"
    ++ (if null (errorDetail e) then "" else ": " ++ errorDetail e)

-- | An attached dongle, as listed by 'devices'.
data Device = Device
  { deviceSerial :: String   -- ^ the serial, as hex
  , devicePath   :: String   -- ^ the device path 'openPath' takes
  } deriving (Eq, Show)

-- | An open dongle. The library holds up to 32 at a time; release them with
-- 'close' or use 'withDongle'.
newtype Dongle = Dongle CInt
  deriving (Eq, Show)

-- | Plaintext device information, read without a session.
data Info = Info
  { protocolVersion    :: (Int, Int)
  , firmwareVersion    :: (Int, Int, Int)
  , secureElementReady :: Bool
  , provisioned        :: Bool
  , watchdogReboot     :: Bool   -- ^ the previous boot ended in a watchdog reset
  , isolated           :: Bool
  , writeAuthRotated   :: Bool   -- ^ 'rotateWriteKey' has replaced the shipped write key
  , dataCapacity       :: Int    -- ^ record storage, bytes
  , dataFree           :: Int
  } deriving (Eq, Show)

-- | What a genuine dongle proves: the serial from its certificate and the day
-- it was personalised (@YYYY-MM-DD@, or empty).
data GenuineResult = GenuineResult
  { genuineSerial   :: String
  , provisionedDate :: String
  } deriving (Eq, Show)

-- | A record on the dongle.
data RecordInfo = RecordInfo
  { recordName :: String
  , recordSize :: Int
  } deriving (Eq, Show)

-- | Who can decrypt data sealed with 'appEncrypt'.
data Scope
  = DeviceScope     -- ^ only the dongle that sealed it
  | DeveloperScope  -- ^ any dongle issued to the same developer
  deriving (Eq, Show)

serialSize, dateSize, pathSize, errorSize :: Int
serialSize = 15
dateSize = 11
pathSize = 512
errorSize = 256

rangeCode :: CInt
rangeCode = -11

-- Failure handling: every flat call returns 0 or a negative status.
check :: Maybe Dongle -> String -> CInt -> IO ()
check h op rc
  | rc == 0 = pure ()
  | otherwise = do
      detail <- maybe (pure "") lastErrorDetail h
      throwIO (LicDongleError (statusFromCode (fromIntegral rc)) (fromIntegral rc) op detail)

withOutInt :: (Ptr CInt -> IO CInt) -> IO (CInt, CInt)
withOutInt f = alloca $ \p -> do
  poke p 0
  rc <- f p
  v <- peek p
  pure (rc, v)

-- A NUL-terminated string into a caller buffer of the given capacity.
readString :: Int -> (CString -> CInt -> IO CInt) -> IO (CInt, String)
readString cap f = allocaBytes cap $ \buf -> do
  pokeByteOff buf 0 (0 :: Word8)
  rc <- f buf (fromIntegral cap)
  s <- if rc == 0 then peekCString buf else pure ""
  pure (rc, s)

-- Bytes of unknown length: ask with a capacity of 0, the library answers
-- Range and the size needed, then read into a buffer of that size.
readBytes :: Maybe Dongle -> String -> (Ptr Word8 -> CInt -> Ptr CInt -> IO CInt) -> IO ByteString
readBytes h op f = do
  (rc0, need) <- withOutInt (f nullPtr 0)
  if rc0 == 0
    then pure BS.empty
    else if rc0 /= rangeCode
      then check h op rc0 >> pure BS.empty
      else allocaBytes (fromIntegral need) $ \buf -> do
        (rc, len) <- withOutInt (f buf need)
        check h op rc
        BS.packCStringLen (castPtr buf, fromIntegral len)

useBytes :: ByteString -> (Ptr Word8 -> CInt -> IO a) -> IO a
useBytes bs f = BS.useAsCStringLen bs $ \(p, n) -> f (castPtr p) (fromIntegral n)

-- | @(major, minor, patch)@ of the loaded native library; loads it if no other
-- call has.
libraryVersion :: IO (Int, Int, Int)
libraryVersion = do
  a <- api
  allocaArray 3 $ \arr -> do
    rc <- fVersion a arr (advancePtr arr 1) (advancePtr arr 2)
    check Nothing "licdf_version" rc
    [x, y, z] <- map fromIntegral <$> peekArray 3 arr
    pure (x, y, z)

-- | The library's text for a status code.
statusText :: Int -> IO String
statusText code = do
  a <- api
  snd <$> readString errorSize (fStrerror a (fromIntegral code))

-- | The library's diagnostic detail for the most recent failure on a dongle.
lastErrorDetail :: Dongle -> IO String
lastErrorDetail (Dongle h) = do
  a <- api
  snd <$> readString errorSize (fLastError a h)

-- | The attached dongles.
devices :: IO [Device]
devices = do
  a <- api
  (rc, n) <- withOutInt (fDeviceCount a)
  check Nothing "licdf_device_count" rc
  forM [0 .. n - 1] $ \i -> do
    (rc1, s) <- readString pathSize (fDeviceSerial a i)
    check Nothing "licdf_device_serial" rc1
    (rc2, p) <- readString pathSize (fDevicePath a i)
    check Nothing "licdf_device_path" rc2
    pure (Device s p)

-- | Opens the first dongle.
open :: IO Dongle
open = openSerial ""

-- | Opens the dongle with that serial (an empty serial means the first one).
openSerial :: String -> IO Dongle
openSerial s = do
  a <- api
  h <- withCString s (fOpen a)
  handleOf "licdf_open" h

-- | Opens the dongle at a device path from 'devices'.
openPath :: String -> IO Dongle
openPath p = do
  a <- api
  h <- withCString p (fOpenPath a)
  handleOf "licdf_open_path" h

handleOf :: String -> CInt -> IO Dongle
handleOf op h
  | h > 0 = pure (Dongle h)
  | otherwise = throwIO (LicDongleError (statusFromCode (fromIntegral h)) (fromIntegral h) op "")

-- | Releases a dongle; its session ends with it.
close :: Dongle -> IO ()
close (Dongle h) = do
  a <- api
  fClose a h >>= check Nothing "licdf_close"

-- | Opens a dongle, runs the action, and closes the dongle on every exit path.
withDongle :: IO Dongle -> (Dongle -> IO a) -> IO a
withDongle acquire = bracket acquire close

-- | The dongle's serial, as hex.
serial :: Dongle -> IO String
serial d@(Dongle h) = do
  a <- api
  (rc, s) <- readString serialSize (fGetSerial a h)
  check (Just d) "licdf_get_serial" rc
  pure s

-- | Plaintext device information.
info :: Dongle -> IO Info
info d@(Dongle h) = do
  a <- api
  allocaArray 8 $ \arr -> do
    let p i = advancePtr arr i
    rc <- fGetInfo a h (p 0) (p 1) (p 2) (p 3) (p 4) (p 5) (p 6) (p 7)
    check (Just d) "licdf_get_info" rc
    v <- map fromIntegral <$> peekArray 8 arr
    let flags = v !! 5 :: Int
        flag b = flags .&. b /= 0
    pure Info
      { protocolVersion = (v !! 0, v !! 1)
      , firmwareVersion = (v !! 2, v !! 3, v !! 4)
      , secureElementReady = flag 0x01
      , provisioned = flag 0x02
      , watchdogReboot = flag 0x04
      , isolated = flag 0x08
      , writeAuthRotated = flag 0x10
      , dataCapacity = v !! 6
      , dataFree = v !! 7
      }

-- | Proves that the dongle is genuine: certificate chain to the trusted root
-- and a live challenge-response. Throws unless it is.
verifyGenuine :: Dongle -> IO GenuineResult
verifyGenuine d@(Dongle h) = do
  a <- api
  alloca $ \pg -> allocaBytes serialSize $ \ps -> allocaBytes dateSize $ \pd -> do
    poke pg 0
    pokeByteOff ps 0 (0 :: Word8)
    pokeByteOff pd 0 (0 :: Word8)
    rc <- fVerifyGenuine a h pg ps (fromIntegral serialSize) pd (fromIntegral dateSize)
    check (Just d) "licdf_verify_genuine" rc
    g <- peek pg
    when (g == 0) $ throwIO (LicDongleError NotGenuine (-7) "licdf_verify_genuine" "")
    GenuineResult <$> peekCString ps <*> peekCString pd

-- | 'True' only when the dongle proves genuine. Every failure, of any kind,
-- gives 'False'.
isGenuine :: Dongle -> IO Bool
isGenuine d = (verifyGenuine d >> pure True) `catch` \(_ :: SomeException) -> pure False

-- | Replaces the CA root that 'verifyGenuine' checks against (a certificate in
-- DER form). Applications do not need this.
setTrustRoot :: Dongle -> ByteString -> IO ()
setTrustRoot d@(Dongle h) der = do
  a <- api
  useBytes der (fSetTrustRoot a h) >>= check (Just d) "licdf_set_trust_root"

-- | Opens the encrypted session that records, counters and application-data
-- encryption need.
sessionOpen :: Dongle -> IO ()
sessionOpen d@(Dongle h) = do
  a <- api
  fSessionOpen a h >>= check (Just d) "licdf_session_open"

-- | Ends the session; the write role ends with it.
sessionClose :: Dongle -> IO ()
sessionClose d@(Dongle h) = do
  a <- api
  fSessionClose a h >>= check (Just d) "licdf_session_close"

-- | Runs the action inside a session, closing it on every exit path.
withSession :: Dongle -> IO a -> IO a
withSession d = bracket_ (sessionOpen d) (sessionClose d)

-- | Elevates the session to the write role with the dongle's write key, a
-- P-256 private key in PKCS#8 DER form.
authorizeWrite :: Dongle -> ByteString -> IO ()
authorizeWrite d@(Dongle h) key = do
  a <- api
  useBytes key (fWriteAuth a h) >>= check (Just d) "licdf_write_auth"

-- | Replaces the dongle's write key with the given one (P-256, PKCS#8 DER).
-- From the next session on, only that key elevates. Needs the write role.
rotateWriteKey :: Dongle -> ByteString -> IO ()
rotateWriteKey d@(Dongle h) key = do
  a <- api
  useBytes key (fWriteAuthRotate a h) >>= check (Just d) "licdf_write_auth_rotate"

-- | The records on the dongle. Needs a session.
records :: Dongle -> IO [RecordInfo]
records d@(Dongle h) = do
  a <- api
  (rc, n) <- withOutInt (fRecordCount a h)
  check (Just d) "licdf_record_count" rc
  forM [0 .. n - 1] $ \i -> allocaBytes pathSize $ \buf -> alloca $ \psz -> do
    pokeByteOff buf 0 (0 :: Word8)
    poke psz 0
    rc1 <- fRecordName a h i buf (fromIntegral pathSize) psz
    check (Just d) "licdf_record_name" rc1
    RecordInfo <$> peekCString buf <*> (fromIntegral <$> peek psz)

-- | Reads a record. Needs a session.
readRecord :: Dongle -> String -> IO ByteString
readRecord d@(Dongle h) name = do
  a <- api
  withCString name $ \cn -> readBytes (Just d) "licdf_record_read" (fRecordRead a h cn)

-- | Replaces (or creates) a record. Needs the write role.
writeRecord :: Dongle -> String -> ByteString -> IO ()
writeRecord d@(Dongle h) name bytes = do
  a <- api
  withCString name $ \cn -> useBytes bytes (fRecordWrite a h cn) >>= check (Just d) "licdf_record_write"

-- | Erases one record. Needs the write role. An empty name is refused; use
-- 'eraseAllRecords' to erase everything.
eraseRecord :: Dongle -> String -> IO ()
eraseRecord d@(Dongle h) name = do
  a <- api
  withCString name (fRecordErase a h) >>= check (Just d) "licdf_record_erase"

-- | Erases every record. Needs the write role.
eraseAllRecords :: Dongle -> IO ()
eraseAllRecords d@(Dongle h) = do
  a <- api
  fRecordEraseAll a h >>= check (Just d) "licdf_record_erase_all"

-- | Reads a hardware monotonic counter (0 or 1). Needs a session.
readCounter :: Dongle -> Int -> IO Int
readCounter d@(Dongle h) i = do
  a <- api
  (rc, v) <- withOutInt (fCounterRead a h (fromIntegral i))
  check (Just d) "licdf_counter_read" rc
  pure (fromIntegral v)

-- | Increments a counter, irreversibly, and gives the new value. Needs the
-- write role.
incrementCounter :: Dongle -> Int -> IO Int
incrementCounter d@(Dongle h) i = do
  a <- api
  (rc, v) <- withOutInt (fCounterIncrement a h (fromIntegral i))
  check (Just d) "licdf_counter_increment" rc
  pure (fromIntegral v)

-- | Seals data so that only a dongle can open it. Needs a session. This is the
-- pair to build a licence check on: route something the program needs through
-- it, so that removing the check removes the data.
appEncrypt :: Dongle -> Scope -> ByteString -> IO ByteString
appEncrypt d@(Dongle h) scope plain = do
  a <- api
  let code = case scope of DeviceScope -> 0; DeveloperScope -> 1
  useBytes plain $ \pp n -> readBytes (Just d) "licdf_app_encrypt" (fAppEncrypt a h code pp n)

-- | Opens data sealed with 'appEncrypt'. Needs a session.
appDecrypt :: Dongle -> ByteString -> IO ByteString
appDecrypt d@(Dongle h) packed = do
  a <- api
  useBytes packed $ \pp n -> readBytes (Just d) "licdf_app_decrypt" (fAppDecrypt a h pp n)

File: bindings/elixir/lib/keynub/lic_dongle.ex
defmodule KeyNub.LicDongle do
  @moduledoc """
  Client for the KeyNub USB license dongle.

      {:ok, secret} =
        KeyNub.LicDongle.with_dongle(fn d ->
          KeyNub.LicDongle.verify_genuine!(d)
          KeyNub.LicDongle.with_session!(d, fn -> KeyNub.LicDongle.app_decrypt!(d, sealed) end)
        end)

  Every call returns `{:ok, value}`, `:ok` or `{:error, %KeyNub.LicDongle.Error{}}`;
  the `!` variants return the value and raise the error instead. The package
  calls the SDK's flat companion API through a small NIF that loads the native
  library at run time; see `KeyNub.LicDongle.Library` for where that library
  comes from.
  """

  alias KeyNub.LicDongle.{Error, Library, Nif}

  defmodule Dongle do
    @moduledoc "An open dongle, from `KeyNub.LicDongle.open/1`."
    @enforce_keys [:handle]
    defstruct [:handle]
    @type t :: %__MODULE__{handle: pos_integer}
  end

  defmodule Device do
    @moduledoc "An attached dongle, from `KeyNub.LicDongle.devices/0`."
    defstruct [:serial, :path]
    @type t :: %__MODULE__{serial: String.t(), path: String.t()}
  end

  defmodule Info do
    @moduledoc "Plaintext device information, from `KeyNub.LicDongle.info/1`."
    defstruct [
      :protocol_version,
      :firmware_version,
      :secure_element_ready,
      :provisioned,
      :watchdog_reboot,
      :isolated,
      :write_auth_rotated,
      :data_capacity,
      :data_free
    ]

    @type t :: %__MODULE__{
            protocol_version: {non_neg_integer, non_neg_integer},
            firmware_version: {non_neg_integer, non_neg_integer, non_neg_integer},
            secure_element_ready: boolean,
            provisioned: boolean,
            watchdog_reboot: boolean,
            isolated: boolean,
            write_auth_rotated: boolean,
            data_capacity: non_neg_integer,
            data_free: non_neg_integer
          }
  end

  defmodule Genuine do
    @moduledoc "The result of a successful `KeyNub.LicDongle.verify_genuine/1`."
    defstruct [:serial, :provisioned_date]
    @type t :: %__MODULE__{serial: String.t(), provisioned_date: String.t()}
  end

  defmodule Record do
    @moduledoc "A record on the dongle, from `KeyNub.LicDongle.records/1`."
    defstruct [:name, :size]
    @type t :: %__MODULE__{name: String.t(), size: non_neg_integer}
  end

  @typedoc "`:device` seals to this dongle only; `:developer` to any dongle issued by the same developer."
  @type scope :: :device | :developer

  @type result(value) :: {:ok, value} | {:error, Error.t()}
  @type status_result :: :ok | {:error, Error.t()}

  @range -11
  @not_genuine -7

  # ---- the library -----------------------------------------------------------

  @doc "Names the native library file to load. See `KeyNub.LicDongle.Library`."
  defdelegate set_library_path(path), to: Library

  @doc "The library path in use, or the first candidate when nothing is loaded yet."
  defdelegate library_path(), to: Library

  @doc "The path of the loaded library, or `nil` before the first call."
  defdelegate loaded_library_path(), to: Library

  @doc "The native library's version as `{major, minor, patch}`."
  @spec library_version() :: result({integer, integer, integer})
  def library_version do
    Library.ensure_loaded!()
    result("licdf_version", Nif.version())
  end

  @doc "Human-readable text for a status code."
  @spec status_text(integer) :: String.t()
  def status_text(code) when is_integer(code) do
    Library.ensure_loaded!()
    Nif.strerror(code)
  end

  @doc "Diagnostic detail for the most recent failure on this dongle (may be empty)."
  @spec last_error_detail(Dongle.t()) :: String.t()
  def last_error_detail(%Dongle{handle: h}) do
    Library.ensure_loaded!()
    Nif.last_error(h)
  end

  # ---- discovery, open, close --------------------------------------------------

  @doc "The attached dongles."
  @spec devices() :: result([Device.t()])
  def devices do
    Library.ensure_loaded!()

    with {:ok, n} <- result("licdf_device_count", Nif.device_count()) do
      map_ok(indices(n), fn i ->
        with {:ok, serial} <- result("licdf_device_serial", Nif.device_serial(i)),
             {:ok, path} <- result("licdf_device_path", Nif.device_path(i)) do
          {:ok, %Device{serial: serial, path: path}}
        end
      end)
    end
  end

  @doc "Opens the dongle with this serial, or the first one found when `nil`."
  @spec open(String.t() | nil) :: result(Dongle.t())
  def open(serial \\ nil) when is_binary(serial) or is_nil(serial) do
    Library.ensure_loaded!()
    handle("licdf_open", Nif.open(serial || ""))
  end

  @doc "Opens the dongle at this device path (from `devices/0`)."
  @spec open_path(String.t()) :: result(Dongle.t())
  def open_path(path) when is_binary(path) do
    Library.ensure_loaded!()
    handle("licdf_open_path", Nif.open_path(path))
  end

  @doc "Closes the dongle. Every further call on it fails."
  @spec close(Dongle.t()) :: status_result
  def close(%Dongle{handle: h}), do: result("licdf_close", Nif.close(h))

  @doc """
  Opens a dongle (by serial, or the first one when `nil`), runs `fun` with it
  and closes it on every exit path. Returns `{:ok, fun_result}`, or the error
  from opening.
  """
  @spec with_dongle(String.t() | nil, (Dongle.t() -> value)) :: result(value) when value: term
  def with_dongle(serial \\ nil, fun) when is_function(fun, 1) do
    case open(serial) do
      {:ok, d} ->
        try do
          {:ok, fun.(d)}
        after
          close(d)
        end

      {:error, _} = e ->
        e
    end
  end

  # ---- plaintext info ------------------------------------------------------------

  @doc "The dongle's serial number (14 hex digits)."
  @spec serial(Dongle.t()) :: result(String.t())
  def serial(%Dongle{handle: h}), do: result("licdf_get_serial", h, Nif.get_serial(h))

  @doc "Plaintext device information."
  @spec info(Dongle.t()) :: result(Info.t())
  def info(%Dongle{handle: h}) do
    with {:ok, {pa, pb, fa, fb, fc, flags, capacity, free}} <-
           result("licdf_get_info", h, Nif.get_info(h)) do
      flag = fn bit -> Bitwise.band(flags, bit) != 0 end

      {:ok,
       %Info{
         protocol_version: {pa, pb},
         firmware_version: {fa, fb, fc},
         secure_element_ready: flag.(0x01),
         provisioned: flag.(0x02),
         watchdog_reboot: flag.(0x04),
         isolated: flag.(0x08),
         write_auth_rotated: flag.(0x10),
         data_capacity: capacity,
         data_free: free
       }}
    end
  end

  @doc """
  Proves the dongle is genuine: certificate chain to the trusted root plus a
  live challenge-response. `{:ok, %Genuine{}}` only when it is.
  """
  @spec verify_genuine(Dongle.t()) :: result(Genuine.t())
  def verify_genuine(%Dongle{handle: h}) do
    case Nif.verify_genuine(h) do
      {0, genuine, serial, date} when genuine != 0 ->
        {:ok, %Genuine{serial: serial, provisioned_date: date}}

      {0, _, _, _} ->
        {:error, Error.new("licdf_verify_genuine", @not_genuine)}

      {rc, _, _, _} ->
        {:error, error("licdf_verify_genuine", h, rc)}
    end
  end

  @doc "The non-raising gate: `true` only when `verify_genuine/1` succeeds. Fails closed."
  @spec genuine?(Dongle.t()) :: boolean
  def genuine?(%Dongle{} = d) do
    match?({:ok, _}, verify_genuine(d))
  rescue
    _ -> false
  end

  @doc "Overrides the CA root that `verify_genuine/1` checks against (DER)."
  @spec set_trust_root(Dongle.t(), binary) :: status_result
  def set_trust_root(%Dongle{handle: h}, der) when is_binary(der) do
    result("licdf_set_trust_root", h, Nif.set_trust_root(h, der))
  end

  # ---- session -------------------------------------------------------------------

  @doc "Opens an authenticated session; records, counters and app crypto need one."
  @spec session_open(Dongle.t()) :: status_result
  def session_open(%Dongle{handle: h}), do: result("licdf_session_open", h, Nif.session_open(h))

  @doc "Closes the session."
  @spec session_close(Dongle.t()) :: status_result
  def session_close(%Dongle{handle: h}),
    do: result("licdf_session_close", h, Nif.session_close(h))

  @doc "Opens a session, runs `fun` and closes the session on every exit path."
  @spec with_session(Dongle.t(), (-> value)) :: result(value) when value: term
  def with_session(%Dongle{} = d, fun) when is_function(fun, 0) do
    case session_open(d) do
      :ok ->
        try do
          {:ok, fun.()}
        after
          session_close(d)
        end

      {:error, _} = e ->
        e
    end
  end

  @doc """
  Elevates the session to the write role with a write-auth key (P-256 PKCS#8
  DER). Belongs in licence-issuing tooling, never in the application users run.
  """
  @spec authorize_write(Dongle.t(), binary) :: status_result
  def authorize_write(%Dongle{handle: h}, key) when is_binary(key) do
    result("licdf_write_auth", h, Nif.write_auth(h, key))
  end

  @doc """
  Replaces the dongle's write-auth key with `key` (P-256 PKCS#8 DER). Call
  `authorize_write/2` first. From the next session on, only the new key elevates.
  """
  @spec rotate_write_key(Dongle.t(), binary) :: status_result
  def rotate_write_key(%Dongle{handle: h}, key) when is_binary(key) do
    result("licdf_write_auth_rotate", h, Nif.write_auth_rotate(h, key))
  end

  # ---- records -------------------------------------------------------------------

  @doc "The records on the dongle, with their sizes."
  @spec records(Dongle.t()) :: result([Record.t()])
  def records(%Dongle{handle: h}) do
    with {:ok, n} <- result("licdf_record_count", h, Nif.record_count(h)) do
      map_ok(indices(n), fn i ->
        case Nif.record_name(h, i) do
          {0, name, size} -> {:ok, %Record{name: name, size: size}}
          {rc, _, _} -> {:error, error("licdf_record_name", h, rc)}
        end
      end)
    end
  end

  @doc "The record's bytes."
  @spec read_record(Dongle.t(), String.t()) :: result(binary)
  def read_record(%Dongle{handle: h}, name) when is_binary(name) do
    read_bytes("licdf_record_read", h, &Nif.record_read(h, name, &1))
  end

  @doc "Writes (creates or replaces) the record. Needs the write role."
  @spec write_record(Dongle.t(), String.t(), binary) :: status_result
  def write_record(%Dongle{handle: h}, name, data) when is_binary(name) and is_binary(data) do
    result("licdf_record_write", h, Nif.record_write(h, name, data))
  end

  @doc "Erases one record by name. An empty name is refused."
  @spec erase_record(Dongle.t(), String.t()) :: status_result
  def erase_record(%Dongle{handle: h}, name) when is_binary(name) do
    result("licdf_record_erase", h, Nif.record_erase(h, name))
  end

  @doc "Erases every record."
  @spec erase_all_records(Dongle.t()) :: status_result
  def erase_all_records(%Dongle{handle: h}) do
    result("licdf_record_erase_all", h, Nif.record_erase_all(h))
  end

  # ---- counters ------------------------------------------------------------------

  @doc "The value of a hardware monotonic counter."
  @spec read_counter(Dongle.t(), non_neg_integer) :: result(non_neg_integer)
  def read_counter(%Dongle{handle: h}, id) when is_integer(id) do
    result("licdf_counter_read", h, Nif.counter_read(h, id))
  end

  @doc "Increments a counter and returns its new value. Needs the write role."
  @spec increment_counter(Dongle.t(), non_neg_integer) :: result(non_neg_integer)
  def increment_counter(%Dongle{handle: h}, id) when is_integer(id) do
    result("licdf_counter_increment", h, Nif.counter_increment(h, id))
  end

  # ---- app-data envelope encryption ------------------------------------------

  @doc """
  Seals `plaintext` so that only a dongle can open it: this one (`:device`) or
  any dongle issued by the same developer (`:developer`). Build the licence
  check on this pair: route data the program needs through it.
  """
  @spec app_encrypt(Dongle.t(), scope, binary) :: result(binary)
  def app_encrypt(%Dongle{handle: h}, scope, plaintext)
      when scope in [:device, :developer] and is_binary(plaintext) do
    code = if scope == :device, do: 0, else: 1
    read_bytes("licdf_app_encrypt", h, &Nif.app_encrypt(h, code, plaintext, &1))
  end

  @doc "Opens data sealed with `app_encrypt/3`."
  @spec app_decrypt(Dongle.t(), binary) :: result(binary)
  def app_decrypt(%Dongle{handle: h}, packed) when is_binary(packed) do
    read_bytes("licdf_app_decrypt", h, &Nif.app_decrypt(h, packed, &1))
  end

  # ---- the raising variants --------------------------------------------------------

  for {name, arity} <- [
        library_version: 0,
        devices: 0,
        open: 0,
        open: 1,
        open_path: 1,
        close: 1,
        with_dongle: 1,
        with_dongle: 2,
        serial: 1,
        info: 1,
        verify_genuine: 1,
        set_trust_root: 2,
        session_open: 1,
        session_close: 1,
        with_session: 2,
        authorize_write: 2,
        rotate_write_key: 2,
        records: 1,
        read_record: 2,
        write_record: 3,
        erase_record: 2,
        erase_all_records: 1,
        read_counter: 2,
        increment_counter: 2,
        app_encrypt: 3,
        app_decrypt: 2
      ] do
    args = Macro.generate_arguments(arity, __MODULE__)

    @doc "Like `#{name}/#{arity}`, returning the value (or `:ok`) and raising `KeyNub.LicDongle.Error` on failure."
    def unquote(:"#{name}!")(unquote_splicing(args)) do
      unwrap!(unquote(name)(unquote_splicing(args)))
    end
  end

  defp unwrap!(:ok), do: :ok
  defp unwrap!({:ok, value}), do: value
  defp unwrap!({:error, %Error{} = e}), do: raise(e)

  # ---- helpers -------------------------------------------------------------------

  defp indices(n) when n <= 0, do: []
  defp indices(n), do: Enum.to_list(0..(n - 1))

  defp map_ok(items, fun) do
    Enum.reduce_while(items, {:ok, []}, fn item, {:ok, acc} ->
      case fun.(item) do
        {:ok, value} -> {:cont, {:ok, [value | acc]}}
        {:error, _} = e -> {:halt, e}
      end
    end)
    |> case do
      {:ok, acc} -> {:ok, Enum.reverse(acc)}
      e -> e
    end
  end

  defp error(operation, h, rc), do: Error.new(operation, rc, Nif.last_error(h))

  defp handle(_operation, h) when h > 0, do: {:ok, %Dongle{handle: h}}
  defp handle(operation, rc), do: {:error, Error.new(operation, rc)}

  # A raw status, or {status, value}, without a handle for the detail text.
  defp result(_operation, 0), do: :ok
  defp result(operation, rc) when is_integer(rc), do: {:error, Error.new(operation, rc)}
  defp result(_operation, {0, value}), do: {:ok, value}
  defp result(operation, {rc, _}), do: {:error, Error.new(operation, rc)}

  # The same, with the dongle's detail text on failure.
  defp result(_operation, _h, 0), do: :ok
  defp result(operation, h, rc) when is_integer(rc), do: {:error, error(operation, h, rc)}
  defp result(_operation, _h, {0, value}), do: {:ok, value}
  defp result(operation, h, {rc, _}), do: {:error, error(operation, h, rc)}

  # Bytes of unknown length: ask with a capacity of 0, the library answers
  # :range and the size needed, then read into that size.
  defp read_bytes(operation, h, call) do
    case call.(0) do
      {0, bytes} ->
        {:ok, bytes}

      {@range, needed} ->
        case call.(needed) do
          {0, bytes} -> {:ok, bytes}
          {rc, _} -> {:error, error(operation, h, rc)}
        end

      {rc, _} ->
        {:error, error(operation, h, rc)}
    end
  end
end
